agent/engine/daemonmanager/daemon_manager_linux.go

```diff
@@ -0,0 +1,166 @@
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package daemonmanager
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/fs"
+	"os"
+
+	"github.com/pborman/uuid"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	apicontainerstatus "github.com/aws/amazon-ecs-agent/agent/api/container/status"
+	apitask "github.com/aws/amazon-ecs-agent/agent/api/task"
+	apitaskstatus "github.com/aws/amazon-ecs-agent/agent/api/task/status"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi"
+	"github.com/aws/amazon-ecs-agent/agent/taskresource"
+	"github.com/aws/amazon-ecs-agent/agent/utils/loader"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	md "github.com/aws/amazon-ecs-agent/ecs-agent/manageddaemon"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/docker/docker/api/types"
+	dockercontainer "github.com/docker/docker/api/types/container"
+)
+
+const (
+	// all daemons will share the same user id
+	// note: AppNetUID is 20000
+	daemonUID                       = 20001
+	ecsAgentLogFileENV              = "ECS_LOGFILE"
+	defaultECSAgentLogPathContainer = "/log"
+)
+
+// each daemon manager manages one single daemon container
+type daemonManager struct {
+	managedDaemon *md.ManagedDaemon
+}
+
+func NewDaemonManager(manageddaemon *md.ManagedDaemon) DaemonManager {
+	return &daemonManager{managedDaemon: manageddaemon}
+}
+
+func (dm *daemonManager) CreateDaemonTask() (*apitask.Task, error) {
+	imageName := dm.managedDaemon.GetLoadedDaemonImageRef()
+	containerRunning := apicontainerstatus.ContainerRunning
+	dockerHostConfig := dockercontainer.HostConfig{
+		NetworkMode: apitask.HostNetworkMode,
+		// the default value of 0 for MaximumRetryCount means retry indefinitely
+		RestartPolicy: dockercontainer.RestartPolicy{
+			Name:              "on-failure",
+			MaximumRetryCount: 0,
+		},
+	}
+	if !dm.managedDaemon.IsValidManagedDaemon() {
+		return nil, fmt.Errorf("%s is an invalid managed daemon", dm.managedDaemon.GetImageName())
+	}
+	for _, mount := range dm.managedDaemon.GetMountPoints() {
+		err := mkdirAllAndChown(mount.SourceVolumeHostPath, 0700, daemonUID, os.Getegid())
+		if err != nil {
+			return nil, err
+		}
+		dockerHostConfig.Binds = append(dockerHostConfig.Binds,
+			fmt.Sprintf("%s:%s", mount.SourceVolumeHostPath, mount.ContainerPath))
+	}
+	rawHostConfig, err := json.Marshal(&dockerHostConfig)
+	if err != nil {
+		return nil, err
+	}
+	healthConfig := dm.managedDaemon.GetDockerHealthConfig()
+	rawHealthConfig, err := json.Marshal(&healthConfig)
+	if err != nil {
+		return nil, err
+	}
+	// The raw host config needs to be created this way - if we marshal the entire config object
```

Comment:
"The raw ~host~ docker config"? since this is not the host config but rather the "docker config" right? I see HostConfig defined above on L80

Reply:
there's Config and HostConfig in the DockerConfig and each is a separate string.
```
			DockerConfig: apicontainer.DockerConfig{
				Config:     aws.String(rawConfig),
				HostConfig: aws.String(string(rawHostConfig)),
			},
```

Reply:
right, I was just referring to the // comment being for `rawConfig` and not `rawHostConfig`. this is non-blocking.